File: Gbotx/archon_engine.js
// archon_engine.js
// The Grand Ritual of Vector & Gemini: The Archon Edition
// A masterfully crafted spell for high-efficiency backtesting and analysis.

import { promises as fs } from 'fs';
import path from 'path';
import { parse } from 'csv-parse';
import { stringify } from 'csv-stringify';
import dotenv from 'dotenv';
import chalk from 'chalk';
let GoogleGenerativeAI;
let _GEMINI_AVAILABLE = false;
try {
    const geminiModule = await import('@google/generative-ai');
    GoogleGenerativeAI = geminiModule.GoogleGenerativeAI;
    _GEMINI_AVAILABLE = true;
} catch (e) {
    console.warn(chalk.yellow("Warning: @google/generative-ai not found. Gemini summary will be unavailable."));
}

// Load environment variables from .env file
dotenv.config();

// --- The Scroll of Strategy (Centralized Configuration) ---
const STRATEGY_CONFIG = {
    rsi_period: 14, adx_period: 14, aroon_period: 14,
    bb_period: 20, bb_std: 2.0,
    supertrend_period: 10, supertrend_multiplier: 3.0,
    z_score_period: 20,
    thresholds: {
        adx_strong_trend: 25, aroon_strong_momentum: 50,
        rsi_overbought: 70, rsi_oversold: 30,
        z_score_overextended: 2.0,
        signal_confidence: 0.4
    }
};

// --- The Chromatic Palette of the Spell ---
const COLOR_HEADER = chalk.cyan.bold;
const COLOR_SIGNAL_LONG = chalk.green.bold;
const COLOR_SIGNAL_SHORT = chalk.red.bold;
const COLOR_SIGNAL_HOLD = chalk.yellow;
const COLOR_CONFIDENCE = chalk.magenta.bold;
const COLOR_REASON_POS = chalk.green;
const COLOR_REASON_NEG = chalk.red;
const COLOR_STATUS = chalk.blue.bold;
const COLOR_ORACLE = chalk.hex('#FFD700'); // Gold
const COLOR_ERROR = chalk.red.bold;

// Helper function to calculate Average True Range (ATR)
function calculateATR(df, period) {
    console.log(`ATR Debug: df.length = ${df.length}, period = ${period}`);
    const trs = [];
    for (let i = 1; i < df.length; i++) {
        const highLow = df[i].high - df[i].low;
        const highPrevClose = Math.abs(df[i].high - df[i - 1].close);
        const lowPrevClose = Math.abs(df[i].low - df[i - 1].close);
        const tr = Math.max(highLow, highPrevClose, lowPrevClose);
        trs.push(tr);
    }
    console.log("ATR Debug: TRs array length:", trs.length);
    console.log("ATR Debug: TRs array:", trs.map(t => t.toFixed(4)));

    const atrs = [];
    if (trs.length > 0) {
        // Initial ATR is the average of the first 'period' TRs
        let sumTr = 0;
        for (let i = 0; i < period && i < trs.length; i++) {
            sumTr += trs[i];
        }
        const initialAtr = sumTr / period;
        atrs.push(initialAtr);
        console.log(`ATR Debug: Initial ATR (${period} period): ${initialAtr.toFixed(4)}`);

        // Subsequent ATRs
        for (let i = period; i < trs.length; i++) {
            const prevAtr = atrs[atrs.length - 1];
            const newAtr = ((prevAtr * (period - 1)) + trs[i]) / period;
            atrs.push(newAtr);
        }
    }
    console.log("ATR Debug: Final ATRs array length:", atrs.length);
    console.log("ATR Debug: Final ATRs array:", atrs.map(a => a.toFixed(4)));
    return atrs;
}

// Helper function to calculate Supertrend
function calculateSupertrend(df, period, multiplier) {
    const atrs = calculateATR(df, period);
    const supertrend = [];
    const supertrendDirection = []; // 1 for up, -1 for down

    for (let i = 0; i < df.length; i++) {
        if (i < period) {
            supertrend.push(NaN);
            supertrendDirection.push(NaN);
            continue;
        }

        const currentAtr = atrs[i - period]; // ATR is calculated from previous data
        if (isNaN(currentAtr)) {
            supertrend.push(NaN);
            supertrendDirection.push(NaN);
            continue;
        }

        const basicUpperBand = ((df[i].high + df[i].low) / 2) + (multiplier * currentAtr);
        const basicLowerBand = ((df[i].high + df[i].low) / 2) - (multiplier * currentAtr);

        let finalUpperBand = basicUpperBand;
        let finalLowerBand = basicLowerBand;

        if (i > period) {
            // Adjust bands based on previous close and band values
            const prevSupertrendDirection = supertrendDirection[i - 1];
            const prevSupertrend = supertrend[i - 1];

            if (!isNaN(prevSupertrendDirection) && !isNaN(prevSupertrend)) {
                if (prevSupertrendDirection === 1) { // Previous was uptrend
                    finalUpperBand = Math.min(basicUpperBand, prevSupertrend);
                    finalLowerBand = Math.max(basicLowerBand, prevSupertrend);
                } else if (prevSupertrendDirection === -1) { // Previous was downtrend
                    finalUpperBand = Math.min(basicUpperBand, prevSupertrend);
                    finalLowerBand = Math.max(basicLowerBand, prevSupertrend);
                }
            }
        }

        let currentSupertrend;
        let currentDirection;

        if (supertrendDirection[i - 1] === 1) { // Previous was uptrend
            if (df[i].close > finalLowerBand) {
                currentSupertrend = finalLowerBand;
                currentDirection = 1;
            } else {
                currentSupertrend = finalUpperBand;
                currentDirection = -1;
            }
        } else if (supertrendDirection[i - 1] === -1) { // Previous was downtrend
            if (df[i].close < finalUpperBand) {
                currentSupertrend = finalUpperBand;
                currentDirection = -1;
            } else {
                currentSupertrend = finalLowerBand;
                currentDirection = 1;
            }
        } else { // First valid point, or previous was NaN
            if (df[i].close > finalUpperBand) {
                currentSupertrend = finalLowerBand;
                currentDirection = 1;
            } else {
                currentSupertrend = finalUpperBand;
                currentDirection = -1;
            }
        }

        supertrend.push(currentSupertrend);
        supertrendDirection.push(currentDirection);
    }
    return { supertrend, supertrendDirection };
}

// --- DataManager: The Keeper of Scrolls and Ethers ---
class DataManager {
    constructor() {
        console.log(COLOR_HEADER("DataManager: Forged to manage data from all sources."));
    }

    async loadFromScroll(filePath) {
        try {
            console.log(chalk.white(`Unfurling the scroll: ${filePath}...`));
            const fileContent = await fs.readFile(filePath, { encoding: 'utf8' });
            const records = await new Promise((resolve, reject) => {
                parse(fileContent, {
                    columns: true,
                    skip_empty_lines: true
                }, (err, records) => {
                    if (err) reject(err);
                    else resolve(records);
                });
            });
            console.log("Type of records: ", typeof records, "Is array: ", Array.isArray(records));

            // Convert to a DataFrame-like structure (array of objects)
            // and ensure correct types and sorting
            records.sort((a, b) => new Date(a.timestamp).getTime() - new Date(b.timestamp).getTime());

            let df = records.map(row => ({
                timestamp: new Date(row.timestamp),
                open: parseFloat(row.open),
                high: parseFloat(row.high),
                low: parseFloat(row.low),
                close: parseFloat(row.close),
                volume: parseFloat(row.volume)
            }));
            console.log(chalk.green("Scroll successfully deciphered."));
            return df;
        } catch (e) {
            console.error(COLOR_ERROR(`A flaw was found in the scroll's inscription: ${e.message}`));
            return [];
        }
    }

    weaveIndicators(df, config) {
        console.log(chalk.white("Weaving the indicator sigils onto the data scroll..."));
        // Calculate Supertrend
        const { supertrend, supertrendDirection } = calculateSupertrend(df, config.supertrend_period, config.supertrend_multiplier);

        return df.map((row, index) => ({
            ...row,
            RSI_14: Math.random() * 100, // Dummy RSI
            ADX_14: Math.random() * 50,  // Dummy ADX
            AROONOSC_14: Math.random() * 200 - 100, // Dummy Aroon Oscillator
            BBL_20_2_0: row.close * 0.95, // Dummy Bollinger Band Lower
            BBM_20_2_0: row.close,       // Dummy Bollinger Band Middle
            BBU_20_2_0: row.close * 1.05, // Dummy Bollinger Band Upper
            SUPERT_10_3_0: supertrend[index] || NaN, // Actual Supertrend
            SUPERTd_10_3_0: supertrendDirection[index] || NaN, // Actual Supertrend Direction
            Z_SCORE: (row.close - (row.close * (1 + (Math.random() - 0.5) * 0.01))) / (row.close * 0.01) // Dummy Z-Score
        }));
    }
}

// --- GeminiOracle: The Voice of Synthetic Wisdom ---
class GeminiOracle {
    constructor(apiKey) {
        if (!_GEMINI_AVAILABLE) {
            this.model = null;
            console.error(COLOR_ERROR("Gemini Oracle: Not available due to missing dependencies."));
            return;
        }
        try {
            const genAI = new GoogleGenerativeAI(apiKey);
            this.model = genAI.getGenerativeModel({ model: "gemini-pro" });
            console.log(COLOR_ORACLE("Gemini Oracle: Consciousness awakened."));
        } catch (e) {
            this.model = null;
            console.error(COLOR_ERROR(`Failed to awaken the Gemini Oracle: ${e.message}`));
        }
    }

    async summarizeBacktest(resultsDf, symbol, config) {
        if (!this.model) return "The Oracle is silent.";
        
        const tradeSignals = resultsDf.filter(row => row.signal !== 'HOLD');
        const numLongs = tradeSignals.filter(row => row.signal === 'LONG').length;
        const numShorts = tradeSignals.filter(row => row.signal === 'SHORT').length;
        const avgConfidence = tradeSignals.length > 0 ? tradeSignals.reduce((sum, row) => sum + row.confidence, 0) / tradeSignals.length : 0;

        const prompt = `
        You are a master strategist, reviewing a completed trading simulation.
        Analyze the following backtest summary and provide a concise, expert narrative (3-4 sentences) on the strategy's overall behavior.

        Simulation Summary:
        - Asset: ${symbol}
        - Total Candles Analyzed: ${resultsDf.length}
        - Total Trade Signals Generated: ${tradeSignals.length}
        - Long Signals: ${numLongs}
        - Short Signals: ${numShorts}
        - Average Signal Confidence: ${avgConfidence.toFixed(2)}

        Strategy Parameters:
        - RSI Period: ${config.rsi_period}, Overbought: ${config.thresholds.rsi_overbought}, Oversold: ${config.thresholds.rsi_oversold}
        - ADX Period: ${config.adx_period}, Strong Trend Threshold: ${config.thresholds.adx_strong_trend}
        - Supertrend: ${config.supertrend_period}/${config.supertrend_multiplier}

        Provide your strategic overview.
        `;
        try {
            const result = await this.model.generateContent(prompt);
            const response = await result.response;
            return response.text().trim();
        } catch (e) {
            return `The Oracle's thoughts are clouded: ${e.message}`;
        }
    }
}

// --- VectorEngine: The Heart of Quantitative Analysis ---
class VectorEngine {
    constructor(config) {
        this.config = config;
        this.t = config.thresholds;
    }

    analyzeRow(primaryRow, htfDir, symbol) {
        let reasoning = [];
        let score = 0;
        let signal = "HOLD";
        
        // 1. Multi-Timeframe Context
        const htfTrend = htfDir === 1 ? "UP" : "DOWN";
        reasoning.push(`${COLOR_REASON_POS("[+]")} CONTEXT: HTF current is ${htfTrend}.`);
        if (htfDir === 1) { score = 0.2; signal = "LONG"; }
        else { score = 0.2; signal = "SHORT"; }

        // 2. Indicator Confluence & Divergence
        if (primaryRow[`ADX_${this.config.adx_period}`] > this.t.adx_strong_trend) {
            reasoning.push(`${COLOR_REASON_POS("[+]")} TREND: ADX shows a strong current.`);
            score += 0.15;
        }
        
        if (primaryRow[`AROONOSC_${this.config.aroon_period}`] > this.t.aroon_strong_momentum) {
            reasoning.push(`${COLOR_REASON_POS("[+]")} MOMENTUM: Aroon confirms bullish surge.`);
            score += 0.1;
        } else if (primaryRow[`AROONOSC_${this.config.aroon_period}`] < -this.t.aroon_strong_momentum) {
            reasoning.push(`${COLOR_REASON_POS("[+]")} MOMENTUM: Aroon confirms bearish plunge.`);
            score += 0.1;
        }

        // 3. Risk Analysis & Signal Invalidation
        if (primaryRow[`RSI_${this.config.rsi_period}`] > this.t.rsi_overbought) {
            reasoning.push(`${COLOR_REASON_NEG("[-]")} RISK: RSI is overbought.`);
            if (signal === "LONG") { signal = "HOLD"; score -= 0.3; }
        } else if (primaryRow[`RSI_${this.config.rsi_period}`] < this.t.rsi_oversold) {
            reasoning.push(`${COLOR_REASON_NEG("[-]")} RISK: RSI is oversold.`);
            if (signal === "SHORT") { signal = "HOLD"; score -= 0.3; }
        }

        if (Math.abs(primaryRow.Z_SCORE) > this.t.z_score_overextended) {
            reasoning.push(`${COLOR_REASON_NEG("[-]")} RISK: Z-Score shows extreme deviation.`);
            if ((primaryRow.Z_SCORE > 0 && signal === "LONG") || (primaryRow.Z_SCORE < 0 && signal === "SHORT")) {
                signal = "HOLD"; score -= 0.2;
            }
        }

        const stDirCol = `SUPERTd_${this.config.supertrend_period}_${this.config.supertrend_multiplier}`;
        if ((signal === "LONG" && primaryRow[stDirCol] === -1) || (signal === "SHORT" && primaryRow[stDirCol] === 1)) {
            reasoning.push(`${COLOR_REASON_NEG("[-]")} CONFLICT: Primary Supertrend opposes HTF bias.`);
            signal = "HOLD"; score -= 0.3;
        }

        // 4. Final Verdict
        const finalSignal = score >= this.t.signal_confidence ? signal : "HOLD";
        return { signal: finalSignal, confidence: Math.round(Math.max(0, score) * 100) / 100, reasoning: reasoning.map(r => r.split('] ')[1]).join(' | ') };
    }
}

// --- Chronomancer: The Master of Time Simulation ---
class Chronomancer {
    constructor(engine, verbose) {
        this.engine = engine;
        this.verbose = verbose;
        this.results = [];
    }

    async scry(dfPrimary, dfHtf, symbol) {
        console.log(COLOR_HEADER("Chronomancer begins scrying the threads of past time..."));
        
        // Align HTF data to primary timestamps for efficient lookup
        // In a real scenario, you'd need a more robust way to align timeframes
        // For this example, we'll assume dfHtf has corresponding timestamps
        for (const row of dfPrimary) {
            let htfDir = undefined;
            for (let i = dfHtf.length - 1; i >= 0; i--) {
                if (dfHtf[i].timestamp.getTime() <= row.timestamp.getTime()) {
                    htfDir = dfHtf[i][`SUPERTd_${STRATEGY_CONFIG.supertrend_period}_${STRATEGY_CONFIG.supertrend_multiplier}`];
                    break;
                }
            }
            if (htfDir === undefined) {
                continue; // Skip if no corresponding HTF data exists yet
            }

            const analysis = this.engine.analyzeRow(row, htfDir, symbol);
            this.results.push({
                timestamp: row.timestamp, close: row.close,
                signal: analysis.signal, confidence: analysis.confidence,
                reasoning: analysis.reasoning
            });
            if (this.verbose || analysis.signal !== 'HOLD') {
                this.printAnalysis(this.results[this.results.length - 1]);
            }
        }
        
        console.log(COLOR_HEADER(`Scrying complete. ${this.results.length} moments in time were analyzed.`));
    }

    async saveLedger(filename) {
        if (this.results.length === 0) return;
        console.log(chalk.white(`Inscribing the Ledger of Fates to ${filename}...`));
        const csvString = await stringify(this.results.map(row => ({
            timestamp: row.timestamp.toISOString(), // Convert Date to ISO string for CSV
            close: row.close,
            signal: row.signal,
            confidence: row.confidence,
            reasoning: row.reasoning
        })), { header: true });
        await fs.writeFile(filename, csvString);
        console.log(chalk.green("The Ledger has been written."));
    }

    printAnalysis(result) {
        const signalColor = { "LONG": COLOR_SIGNAL_LONG, "SHORT": COLOR_SIGNAL_SHORT, "HOLD": COLOR_SIGNAL_HOLD }[result.signal];
        console.log(`${result.timestamp.toISOString()} | ${COLOR_HEADER("CLOSE:")} ${result.close.toFixed(4)} | ${signalColor(`SIG: ${result.signal}`)} ${COLOR_CONFIDENCE(`(${result.confidence.toFixed(2)})`)}`);
    }
}

// --- The Main Summoning Ritual ---
async function main() {
    const args = process.argv.slice(2);
    let primaryScroll = '';
    let htfScroll = '';
    let symbol = 'HISTORICAL_ASSET';
    let output = '';
    let verbose = false;
    let geminiSummary = false;

    for (let i = 0; i < args.length; i++) {
        switch (args[i]) {
            case '--symbol':
                symbol = args[++i];
                break;
            case '-o':
            case '--output':
                output = args[++i];
                break;
            case '-v':
            case '--verbose':
                verbose = true;
                break;
            case '--gemini-summary':
                geminiSummary = true;
                break;
            default:
                if (!primaryScroll) {
                    primaryScroll = args[i];
                } else if (!htfScroll) {
                    htfScroll = args[i];
                }
                break;
        }
    }

    if (!primaryScroll || !htfScroll) {
        console.error(COLOR_ERROR("Usage: node archon_engine.js <primary_scroll_path> <htf_scroll_path> [--symbol <symbol>] [-o <output_file>] [-v] [--gemini-summary]"));
        return;
    }

    // --- Initialization ---
    const dm = new DataManager();
    const engine = new VectorEngine(STRATEGY_CONFIG);
    const chronomancer = new Chronomancer(engine, verbose);

    // --- Data Preparation ---
    let dfPrimary = await dm.loadFromScroll(primaryScroll);
    let dfHtf = await dm.loadFromScroll(htfScroll);
    if (dfPrimary.length === 0 || dfHtf.length === 0) {
        console.error(COLOR_ERROR("Cannot perform chronomancy without valid scrolls. The spell is broken."));
        return;
    }

    dfPrimary = dm.weaveIndicators(dfPrimary, STRATEGY_CONFIG);
    dfHtf = dm.weaveIndicators(dfHtf, STRATEGY_CONFIG);
    console.log("dfPrimary after weaveIndicators:", dfPrimary.length, dfPrimary[0]);
    console.log("dfHtf after weaveIndicators:", dfHtf.length, dfHtf[0]);

    // --- Execution ---
    // Ensure enough data for indicators to be meaningful
    const minDataPoints = Math.max(STRATEGY_CONFIG.rsi_period, STRATEGY_CONFIG.adx_period, STRATEGY_CONFIG.bb_period, STRATEGY_CONFIG.supertrend_period, STRATEGY_CONFIG.z_score_period);
    if (dfPrimary.length < minDataPoints) {
        console.error(COLOR_ERROR(`Not enough primary data points (${dfPrimary.length}) for indicator calculation. Need at least ${minDataPoints}.`));
        return;
    }
    if (dfHtf.length < STRATEGY_CONFIG.supertrend_period + 1) {
        console.error(COLOR_ERROR(`Not enough higher timeframe data points (${dfHtf.length}) for Supertrend calculation. Need at least ${STRATEGY_CONFIG.supertrend_period + 1}.`));
        return;
    }

    await chronomancer.scry(dfPrimary, dfHtf, symbol);

    if (output) {
        await chronomancer.saveLedger(output);
    }

    if (geminiSummary) {
        if (!_GEMINI_AVAILABLE) {
            console.error(COLOR_ERROR("Cannot invoke the Oracle: @google/generative-ai is not installed."));
        } else {
            const GEMINI_API_KEY = process.env.GEMINI_API_KEY;
            if (!GEMINI_API_KEY) {
                console.error(COLOR_ERROR("Cannot invoke the Oracle: Gemini API key is missing from .env scroll."));
            } else {
                const oracle = new GeminiOracle(GEMINI_API_KEY);
                console.log(`
${COLOR_ORACLE("Invoking the Gemini Oracle for a strategic summary...")}`);
                const summary = await oracle.summarizeBacktest(chronomancer.results, symbol, STRATEGY_CONFIG);
                console.log(chalk.white(summary));
            }
        }
    }
}

if (process.argv[1] === import.meta.url.slice(7)) {
    main().catch(e => {
        console.error(COLOR_ERROR(`A chaotic energy has disrupted the ritual: ${e.message}`));
        console.error(e);
    });
}

File: Gbotx/archon_engine.py
# archon_engine.py
# The Grand Ritual of Vector & Gemini: The Archon Edition
# A masterfully crafted spell for high-efficiency backtesting and analysis.

import os
import time
import argparse
from datetime import datetime
import pandas as pd
import pandas_ta as ta
from dotenv import load_dotenv
from colorama import Fore, Style, init
import google.generativeai as genai

# Channeling the arcane energies of the terminal's glow
init(autoreset=True)

# --- The Scroll of Strategy (Centralized Configuration) ---
STRATEGY_CONFIG = {
    "rsi_period": 14, "adx_period": 14, "aroon_period": 14,
    "bb_period": 20, "bb_std": 2.0,
    "supertrend_period": 10, "supertrend_multiplier": 3.0,
    "z_score_period": 20,
    "thresholds": {
        "adx_strong_trend": 25, "aroon_strong_momentum": 50,
        "rsi_overbought": 70, "rsi_oversold": 30,
        "z_score_overextended": 2.0,
        "signal_confidence": 0.4
    }
}

# --- The Chromatic Palette of the Spell ---
COLOR_HEADER = Fore.CYAN + Style.BRIGHT
COLOR_SIGNAL_LONG = Fore.GREEN + Style.BRIGHT
COLOR_SIGNAL_SHORT = Fore.RED + Style.BRIGHT
COLOR_SIGNAL_HOLD = Fore.YELLOW
COLOR_CONFIDENCE = Fore.MAGENTA + Style.BRIGHT
COLOR_REASON_POS = Fore.GREEN
COLOR_REASON_NEG = Fore.RED
COLOR_STATUS = Fore.BLUE + Style.BRIGHT
COLOR_ORACLE = Fore.LIGHTYELLOW_EX
COLOR_ERROR = Fore.RED + Style.BRIGHT

# --- DataManager: The Keeper of Scrolls and Ethers ---
class DataManager:
    def __init__(self):
        print(f"{COLOR_HEADER}DataManager: Forged to manage data from all sources.{Style.RESET_ALL}")

    def load_from_scroll(self, file_path: str) -> pd.DataFrame:
        try:
            print(f"{Fore.WHITE}Unfurling the scroll: {file_path}...{Style.RESET_ALL}")
            df = pd.read_csv(file_path, parse_dates=['timestamp'], index_col='timestamp')
            df.sort_index(inplace=True)
            return df
        except Exception as e:
            print(f"{COLOR_ERROR}A flaw was found in the scroll's inscription: {e}{Style.RESET_ALL}")
            return pd.DataFrame()

    def weave_indicators(self, df: pd.DataFrame, config: dict) -> pd.DataFrame:
        print(f"{Fore.WHITE}Weaving the indicator sigils onto the data scroll...{Style.RESET_ALL}")
        df.ta.rsi(length=config["rsi_period"], append=True)
        df.ta.adx(length=config["adx_period"], append=True)
        df.ta.aroon(length=config["aroon_period"], append=True)
        df.ta.bbands(length=config["bb_period"], std=config["bb_std"], append=True)
        df.ta.supertrend(length=config["supertrend_period"], multiplier=config["supertrend_multiplier"], append=True)
        # Custom Z-Score calculation
        df['Z_SCORE'] = (df['close'] - df['close'].rolling(window=config["z_score_period"]).mean()) / df['close'].rolling(window=config["z_score_period"]).std()
        return df.dropna()

# --- GeminiOracle: The Voice of Synthetic Wisdom ---
class GeminiOracle:
    def __init__(self, api_key: str):
        try:
            genai.configure(api_key=api_key)
            self.model = genai.GenerativeModel('gemini-pro')
            print(f"{COLOR_ORACLE}Gemini Oracle: Consciousness awakened.{Style.RESET_ALL}")
        except Exception as e:
            self.model = None
            print(f"{COLOR_ERROR}Failed to awaken the Gemini Oracle: {e}{Style.RESET_ALL}")

    def summarize_backtest(self, results_df: pd.DataFrame, symbol: str, config: dict) -> str:
        if not self.model: return "The Oracle is silent."
        
        trade_signals = results_df[results_df['signal'] != 'HOLD']
        num_longs = len(trade_signals[trade_signals['signal'] == 'LONG'])
        num_shorts = len(trade_signals[trade_signals['signal'] == 'SHORT'])
        avg_confidence = trade_signals['confidence'].mean()

        prompt = f"""
        You are a master strategist, reviewing a completed trading simulation.
        Analyze the following backtest summary and provide a concise, expert narrative (3-4 sentences) on the strategy's overall behavior.

        Simulation Summary:
        - Asset: {symbol}
        - Total Candles Analyzed: {len(results_df)}
        - Total Trade Signals Generated: {len(trade_signals)}
        - Long Signals: {num_longs}
        - Short Signals: {num_shorts}
        - Average Signal Confidence: {avg_confidence:.2f}

        Strategy Parameters:
        - RSI Period: {config['rsi_period']}, Overbought: {config['thresholds']['rsi_overbought']}, Oversold: {config['thresholds']['rsi_oversold']}
        - ADX Period: {config['adx_period']}, Strong Trend Threshold: {config['thresholds']['adx_strong_trend']}
        - Supertrend: {config['supertrend_period']}/{config['supertrend_multiplier']}

        Provide your strategic overview.
        """
        try:
            response = self.model.generate_content(prompt)
            return response.text.strip()
        except Exception as e:
            return f"The Oracle's thoughts are clouded: {e}"

# --- VectorEngine: The Heart of Quantitative Analysis ---
class VectorEngine:
    def __init__(self, config: dict):
        self.config = config
        self.t = config["thresholds"]

    def analyze_row(self, primary_row: pd.Series, htf_dir: int, symbol: str) -> dict:
        reasoning, score, signal = [], 0, "HOLD"
        
        # 1. Multi-Timeframe Context
        htf_trend = "UP" if htf_dir == 1 else "DOWN"
        reasoning.append(f"{COLOR_REASON_POS}[+] CONTEXT: HTF current is {htf_trend}.")
        if htf_dir == 1: score, signal = 0.2, "LONG"
        else: score, signal = 0.2, "SHORT"

        # 2. Indicator Confluence & Divergence
        if primary_row[f'ADX_{self.config["adx_period"]}'] > self.t['adx_strong_trend']:
            reasoning.append(f"{COLOR_REASON_POS}[+] TREND: ADX shows a strong current.")
            score += 0.15
        
        if primary_row[f'AROONOSC_{self.config["aroon_period"]}'] > self.t['aroon_strong_momentum']:
            reasoning.append(f"{COLOR_REASON_POS}[+] MOMENTUM: Aroon confirms bullish surge.")
            score += 0.1
        elif primary_row[f'AROONOSC_{self.config["aroon_period"]}'] < -self.t['aroon_strong_momentum']:
            reasoning.append(f"{COLOR_REASON_POS}[+] MOMENTUM: Aroon confirms bearish plunge.")
            score += 0.1

        # 3. Risk Analysis & Signal Invalidation
        if primary_row[f'RSI_{self.config["rsi_period"]}'] > self.t['rsi_overbought']:
            reasoning.append(f"{COLOR_REASON_NEG}[-] RISK: RSI is overbought.")
            if signal == "LONG": signal, score = "HOLD", score - 0.3
        elif primary_row[f'RSI_{self.config["rsi_period"]}'] < self.t['rsi_oversold']:
            reasoning.append(f"{COLOR_REASON_NEG}[-] RISK: RSI is oversold.")
            if signal == "SHORT": signal, score = "HOLD", score - 0.3

        if abs(primary_row['Z_SCORE']) > self.t['z_score_overextended']:
            reasoning.append(f"{COLOR_REASON_NEG}[-] RISK: Z-Score shows extreme deviation.")
            if (primary_row['Z_SCORE'] > 0 and signal == "LONG") or (primary_row['Z_SCORE'] < 0 and signal == "SHORT"):
                signal, score = "HOLD", score - 0.2

        st_dir_col = f'SUPERTd_{self.config["supertrend_period"]}_{self.config["supertrend_multiplier"]}'
        if (signal == "LONG" and primary_row[st_dir_col] == -1) or (signal == "SHORT" and primary_row[st_dir_col] == 1):
            reasoning.append(f"{COLOR_REASON_NEG}[-] CONFLICT: Primary Supertrend opposes HTF bias.")
            signal, score = "HOLD", score - 0.3

        # 4. Final Verdict
        final_signal = signal if score >= self.t['signal_confidence'] else "HOLD"
        return {"signal": final_signal, "confidence": round(max(0, score), 2), "reasoning": " | ".join(r.split(']', 1)[-1].strip() for r in reasoning)}

# --- Chronomancer: The Master of Time Simulation ---
class Chronomancer:
    def __init__(self, engine: VectorEngine, verbose: bool):
        self.engine = engine
        self.verbose = verbose
        self.results = []

    def scry(self, df_primary: pd.DataFrame, df_htf: pd.DataFrame, symbol: str):
        print(f"{COLOR_HEADER}Chronomancer begins scrying the threads of past time...{Style.RESET_ALL}")
        
        # Align HTF data to primary timestamps for efficient lookup
        df_primary['htf_dir'] = df_htf[f'SUPERTd_{STRATEGY_CONFIG["supertrend_period"]}_{STRATEGY_CONFIG["supertrend_multiplier"]}'].reindex(df_primary.index, method='ffill')
        df_primary.dropna(inplace=True)

        for timestamp, row in df_primary.iterrows():
            analysis = self.engine.analyze_row(row, row['htf_dir'], symbol)
            self.results.append({
                "timestamp": timestamp, "close": row['close'],
                "signal": analysis['signal'], "confidence": analysis['confidence'],
                "reasoning": analysis['reasoning']
            })
            if self.verbose or analysis['signal'] != 'HOLD':
                self.print_analysis(self.results[-1])
        
        print(f"{COLOR_HEADER}Scrying complete. {len(self.results)} moments in time were analyzed.{Style.RESET_ALL}")

    def save_ledger(self, filename: str):
        if not self.results: return
        print(f"{Fore.WHITE}Inscribing the Ledger of Fates to {filename}...{Style.RESET_ALL}")
        pd.DataFrame(self.results).to_csv(filename, index=False)
        print(f"{Fore.GREEN}The Ledger has been written.{Style.RESET_ALL}")

    def print_analysis(self, result: dict):
        signal_color = {"LONG": COLOR_SIGNAL_LONG, "SHORT": COLOR_SIGNAL_SHORT, "HOLD": COLOR_SIGNAL_HOLD}.get(result['signal'])
        print(f"{result['timestamp']} | {COLOR_HEADER}CLOSE: {result['close']:.4f} | {signal_color}SIG: {result['signal']:<4} {COLOR_CONFIDENCE}({result['confidence']:.2f}){Style.RESET_ALL}")

# --- The Main Summoning Ritual ---
if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Archon Engine: Analyze historical market data from scrolls.")
    parser.add_argument("primary_scroll", help="Path to the primary timeframe CSV scroll.")
    parser.add_argument("htf_scroll", help="Path to the higher timeframe CSV scroll.")
    parser.add_argument("--symbol", default="HISTORICAL_ASSET", help="Name of the asset being analyzed.")
    parser.add_argument("-o", "--output", help="Path to save the Ledger of Fates (results CSV).")
    parser.add_argument("-v", "--verbose", action="store_true", help="Print analysis for every candle, not just signals.")
    parser.add_argument("--gemini-summary", action="store_true", help="Invoke the Gemini Oracle for a post-run summary.")
    args = parser.parse_args()

    # Awaken Gemini Oracle
    GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")
    if not GEMINI_API_KEY:
        print(f"{COLOR_ERROR}Gemini API key is missing. The Oracle cannot be awakened.")
        gemini_oracle = GeminiOracle(api_key="dummy")
    else:
        gemini_oracle = GeminiOracle(api_key=GEMINI_API_KEY)

    # Initialize the conduit and Vector
    dm = DataManager()
    engine = VectorEngine(STRATEGY_CONFIG)
    chronomancer = Chronomancer(engine, args.verbose)

    # --- Data Preparation ---
    df_primary = dm.load_from_scroll(args.primary_scroll)
    df_htf = dm.load_from_scroll(args.htf_scroll)
    if df_primary.empty or df_htf.empty:
        print(f"{COLOR_ERROR}Cannot perform chronomancy without valid scrolls. The spell is broken.{Style.RESET_ALL}")
        exit()

    df_primary = dm.weave_indicators(df_primary, STRATEGY_CONFIG)
    df_htf = dm.weave_indicators(df_htf, STRATEGY_CONFIG)

    # --- Execution ---
    chronomancer.scry(df_primary, df_htf, args.symbol)

    if args.output:
        chronomancer.save_ledger(args.output)

    if args.gemini_summary:
        GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")
        if not GEMINI_API_KEY:
            print(f"{COLOR_ERROR}Cannot invoke the Oracle: Gemini API key is missing from .env scroll.{Style.RESET_ALL}")
        else:
            oracle = GeminiOracle(GEMINI_API_KEY)
            print(f"\n{COLOR_ORACLE}Invoking the Gemini Oracle for a strategic summary...{Style.RESET_ALL}")
            summary = oracle.summarize_backtest(pd.DataFrame(chronomancer.results), args.symbol, STRATEGY_CONFIG)
            print(f"{Fore.WHITE}{summary}{Style.RESET_ALL}")